plugins/flutter_plugins/path_provider_fde/linux/path_provider_plugin.cc

```diff
@@ -0,0 +1,161 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "path_provider_plugin.h"
+
+#include <flutter/method_channel.h>
+#include <flutter/plugin_registrar.h>
+#include <flutter/standard_method_codec.h>
+
+#include <fcntl.h>
+#include <linux/limits.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <codecvt>
+#include <cstring>
+#include <iostream>
+#include <locale>
+#include <memory>
+#include <sstream>
+#include <string>
+namespace {
+
+using flutter::EncodableValue;
+
+// Look here for GNOME based directories:
+// https://askubuntu.com/questions/14535/whats-the-local-folder-for-in-my-home-directory
+
+// Returns the name of the executable, without the .exe extension, or an empty
+// string on failure.
+std::string GetExecutableName() {
+  const size_t bufSize = PATH_MAX + 1;
+  char dirNameBuffer[bufSize];
+  // Read the symbolic link '/proc/self/exe'.
+  const char *linkName = "/proc/self/exe";
+  const int ret = int(readlink(linkName, dirNameBuffer, bufSize - 1));
+  if (ret < 0) {
+    return "";
+  }
+  std::string dirName = std::string(dirNameBuffer);
+  return dirName.substr(dirName.find_last_of("/") + 1);
+}
+
+void EnsureCreated(std::string path) {
+  if (mkdir(path.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH) == -1) {
+    if (errno == EEXIST) {
+      // alredy exists
+    } else {
+      // something else
+      std::cerr << "cannot create " << path << " error:" << strerror(errno)
```

Comment:
This function isn't really ensuring that the folder exist, despite the name, since it can fail silently (from an API perspective). If the API contract for the plugin is that the returned directory must exist, then failure should be propagated as an `Error(...)` call with details.